File: lambda/src/discord.ts
import { DISCORD_ALLOWED_SERVERS, DISCORD_CLIENT_ID, DISCORD_CLIENT_SECRET } from "./const";
import { id7 } from "./utils";

export interface DiscordInfo {
  id: string;
  hash: string;
  name: string | null;
  avatar: string | null;
  guild: string[];
}

class HTTPError extends Error {
  status: number;
  body: any;
  constructor(status: number, body: any) {
    super(typeof body === "string" ? body : JSON.stringify(body));
    this.status = status;
    this.body = body;
  }
}

export async function discord(code: string, redirect: string): Promise<DiscordInfo> {
  try {
    const accessToken = await _auth_discord(code, redirect);
    const guilds = await _check_joined(accessToken);
    const info = await _get_avatar(accessToken);
    info.guild = guilds;
    return info;
  } catch (err) {
    throw new HTTPError(401, err instanceof Error ? err.stack ?? err.message : String(err));
  }
}

async function _auth_discord(code: string, redirect: string): Promise<string> {
  const url = "https://discord.com/api/oauth2/token";
  const params = new URLSearchParams({
    client_id: DISCORD_CLIENT_ID,
    client_secret: DISCORD_CLIENT_SECRET,
    grant_type: "authorization_code",
    code: code,
    redirect_uri: redirect,
    scope: "identify"
  });

  const res = await fetch(url, {
    method: "POST",
    headers: { "Content-Type": "application/x-www-form-urlencoded" },
    body: params.toString()
  });

  let body: any;
  try {
    body = await res.json();
  } catch {
    throw new HTTPError(401, "invalid response from token endpoint");
  }

  if (!body || typeof body.access_token !== "string") {
    throw new HTTPError(401, body ?? "no access_token");
  }
  return body.access_token;
}

async function _get_avatar(accessToken: string): Promise<DiscordInfo> {
  const url = "https://discordapp.com/api/users/@me";
  const res = await fetch(url, {
    method: "GET",
    headers: { Authorization: `Bearer ${accessToken}` }
  });
  const body = await res.json();
  return {
    id: String(body.id),
    hash: id7(String(body.id)),
    name: body.global_name ?? body.username ?? null,
    avatar: body.avatar ?? null,
    guild: []
  };
}

async function _check_joined(accessToken: string): Promise<string[]> {
  const url = "https://discordapp.com/api/users/@me/guilds";
  const res = await fetch(url, {
    method: "GET",
    headers: { Authorization: `Bearer ${accessToken}` }
  });
  const body = await res.json();
  const serverNames: string[] = [];

  const allowed_servers: { [key: string]: string } = Object.fromEntries(
    DISCORD_ALLOWED_SERVERS.split(",").map(item => {
      const [label, id] = item.split(":");
      return [id, label];
    })
  );

  if (Array.isArray(body)) {
    for (const guild of body) {
      if (guild && typeof guild.id === "string" && Object.prototype.hasOwnProperty.call(allowed_servers, guild.id)) {
        serverNames.push(allowed_servers[guild.id]);
      }
    }
  }
  if (serverNames.length > 0) return serverNames;
  throw new HTTPError(401, "server not allowed");
}

File: api/discord.py
from httpx import AsyncClient
from .utils import id7
from asyncio import gather
from fastapi import HTTPException
from .config import DISCORD_ALLOWED_SERVERS, DISCORD_CLIENT_ID, DISCORD_CLIENT_SECRET
from dataclasses import dataclass, field


@dataclass(frozen=True)
class DiscordInfo:
    id: str
    hash: str
    name: str | None
    avatar: str | None
    guild: list[str] = field(default_factory=list)


async def discord(code: str | None = None, redirect: str | None = None) -> DiscordInfo:
    if not code or not redirect:
        raise HTTPException(400, "code and redirect are required")
    async with AsyncClient() as client:
        try:
            access_token = await _auth_discord(client, code, redirect)
            info, guilds = await gather(
                _get_avatar_data(client, access_token),
                _check_joined(client, access_token),
            )
            return DiscordInfo(
                id=info["id"],
                hash=info["hash"],
                name=info["name"],
                avatar=info["avatar"],
                guild=guilds,
            )
        except HTTPException:
            raise
        except Exception as err:
            raise HTTPException(401, str(err))


async def _auth_discord(client: AsyncClient, code: str, redirect: str) -> str:
    url = "https://discord.com/api/oauth2/token"
    data = {
        "client_id": DISCORD_CLIENT_ID,
        "client_secret": DISCORD_CLIENT_SECRET,
        "grant_type": "authorization_code",
        "code": code,
        "redirect_uri": redirect,
        "scope": "identify",
    }
    response = await client.post(url, data=data)
    try:
        body = response.json()
    except Exception:
        raise HTTPException(401, "invalid response from token endpoint")
    if response.is_error or "access_token" not in body:
        raise HTTPException(401, str(body.get("error_description", body)))
    return body["access_token"]


async def _get_avatar_data(client: AsyncClient, access_token: str) -> dict:
    """内部処理用。dataclass生成前の生データを辞書で返す"""
    url = "https://discordapp.com/api/users/@me"
    headers = {"Authorization": f"Bearer {access_token}"}

    response = await client.get(url, headers=headers)
    body = response.json()

    user_id = str(body.get("id"))
    return {
        "id": user_id,
        "hash": id7(user_id),
        "name": body.get("global_name") or body.get("username"),
        "avatar": body.get("avatar"),
    }


async def _check_joined(client: AsyncClient, access_token: str) -> list[str]:
    url = "https://discordapp.com/api/users/@me/guilds"
    headers = {"Authorization": f"Bearer {access_token}"}

    response = await client.get(url, headers=headers)
    body = response.json()

    # Allowed Servers のパース (辞書内包表記)
    allowed_servers: dict[str, str] = {
        s_id: label
        for item in DISCORD_ALLOWED_SERVERS.split(",")
        if ":" in item
        for label, s_id in [item.split(":", 1)]
    }

    server_names: list[str] = []
    if isinstance(body, list):
        for guild in body:
            g_id = str(guild.get("id"))
            if g_id in allowed_servers:
                server_names.append(allowed_servers[g_id])

    if server_names:
        return server_names

    raise HTTPException(401, "server not allowed")
